@turbo/gen@2.8.8 no longer able to import @inquirer/prompts in config.ts

### Verify canary release

- [x] I verified that the issue exists in the latest Turborepo canary release.

### Link to code that reproduces this issue

https://github.com/jahands/turbo-gen-bug-20250214

### Which canary version will you have in your reproduction?

@turbo/gen@2.8.8 (there are no canaries for @turbo/gen at this time)

### Environment information

### Expected behavior

Ability to use input/search etc from @inquirer/prompts library

### Actual behavior

```
$ bunx @turbo/gen@2.8.8

>>> Modify "turbo-gen-bug-20250214" using custom generators

>>> ResolveMessage: Cannot find module '@inquirer/prompts' from '/Users/jh/src/turbo-gen-bug-20250214/turbo/generators/config.ts'
>>> No generators found.

? Would you like to add a config with a sample custom generator to turbo-gen-bug-20250214? (Y/n)
```

### To Reproduce

```sh
git clone https://github.com/jahands/turbo-gen-bug-20250214
cd turbo-gen-bug-20250214
bun install
bunx @turbo/gen@2.8.7 # works
bunx @turbo/gen@2.8.8 # doesn't work
```

### Additional context

Works on 2.8.7, broke on 2.8.8
No canary releases for @turbo/gen at this time to test it on.

--- a/packages/turbo-gen/__tests__/binary.test.ts
+++ b/packages/turbo-gen/__tests__/binary.test.ts
@@ -80,6 +80,54 @@ module.exports = function generator(plop) {
 };
 `;
 
+// Config that imports an external npm package from the project's node_modules.
+const TS_CONFIG_WITH_EXTERNAL_IMPORT = (name: string) => `
+import { helper } from "test-external-pkg";
+
+export default function generator(plop: any): void {
+  plop.setGenerator("${name}", {
+    description: helper("${name}"),
+    prompts: [],
+    actions: [{ type: "add", path: "out/${name}.md", template: "# ${name}" }]
+  });
+}
+`;
+
+// Config that imports @inquirer/prompts, which is bundled in the compiled
+// binary but NOT installed in the user's node_modules.
+// Regression test for https://github.com/vercel/turborepo/issues/11855
+const TS_CONFIG_WITH_BINARY_MODULE_IMPORT = (name: string) => `
+import { Separator } from "@inquirer/prompts";
+
+const sep = new Separator();
+
+export default function generator(plop: any): void {
+  plop.setGenerator("${name}", {
+    description: "${name}",
+    prompts: [],
+    actions: [{ type: "add", path: "out/${name}.md", template: "# ${name}" }]
+  });
+}
+`;
+
+/**
+ * Creates a fake npm package in the project's node_modules so that
+ * config files can import it without running a real package manager.
+ */
+function createFakePackage(
+  projectRoot: string,
+  packageName: string,
+  code: string
+) {
+  const pkgDir = path.join(projectRoot, "node_modules", packageName);
+  fs.mkdirSync(pkgDir, { recursive: true });
+  fs.writeFileSync(
+    path.join(pkgDir, "package.json"),
+    JSON.stringify({ name: packageName, version: "1.0.0", main: "index.js" })
+  );
+  fs.writeFileSync(path.join(pkgDir, "index.js"), code);
+}
+
 // Skip the entire suite if the binary hasn't been built.
 // Unit tests (raw.test.ts) always run; these are integration tests
 // that require `pnpm --filter @turbo/gen run build` first.
@@ -224,6 +272,97 @@ describeIfBinary("compiled binary", () => {
       });
     }
   );
+
+  // Regression test for https://github.com/vercel/turborepo/issues/11855
+  describe("config importing external npm packages", () => {
+    let projectDir: string;
+    const genName = "ext-import-test";
+
+    beforeAll(() => {
+      projectDir = path.join(tmpDir, "external-import");
+      fs.mkdirSync(projectDir, { recursive: true });
+      createProject(projectDir, {
+        type: "commonjs",
+        configFile: "config.ts",
+        configContent: TS_CONFIG_WITH_EXTERNAL_IMPORT(genName),
+        generatorPkgType: "commonjs"
+      });
+      // Install a fake package in node_modules
+      createFakePackage(
+        projectDir,
+        "test-external-pkg",
+        'module.exports.helper = function(name) { return name + " via external"; };'
+      );
+    });
+
+    it("resolves npm packages from the project node_modules", () => {
+      fs.rmSync(path.join(projectDir, "out"), {
+        recursive: true,
+        force: true
+      });
+
+      bin(
+        [
+          "raw",
+          "run",
+          "--json",
+          JSON.stringify({
+            root: projectDir,
+            generator_name: genName
+          })
+        ],
+        projectDir
+      );
+
+      const outFile = path.join(projectDir, "out", `${genName}.md`);
+      expect(fs.existsSync(outFile)).toBe(true);
+      expect(fs.readFileSync(outFile, "utf-8")).toContain(`# ${genName}`);
+    });
+  });
+
+  // Regression test for https://github.com/vercel/turborepo/issues/11855
+  // @inquirer/prompts is bundled in the binary but NOT in the test project's
+  // node_modules, which is the exact scenario from the bug report.
+  describe("config importing binary-bundled modules (@inquirer/prompts)", () => {
+    let projectDir: string;
+    const genName = "binary-module-test";
+
+    beforeAll(() => {
+      projectDir = path.join(tmpDir, "binary-module-import");
+      fs.mkdirSync(projectDir, { recursive: true });
+      createProject(projectDir, {
+        type: "commonjs",
+        configFile: "config.ts",
+        configContent: TS_CONFIG_WITH_BINARY_MODULE_IMPORT(genName),
+        generatorPkgType: "commonjs"
+      });
+      // Intentionally NO node_modules/@inquirer/prompts on disk
+    });
+
+    it("resolves @inquirer/prompts from the compiled binary", () => {
+      fs.rmSync(path.join(projectDir, "out"), {
+        recursive: true,
+        force: true
+      });
+
+      bin(
+        [
+          "raw",
+          "run",
+          "--json",
+          JSON.stringify({
+            root: projectDir,
+            generator_name: genName
+          })
+        ],
+        projectDir
+      );
+
+      const outFile = path.join(projectDir, "out", `${genName}.md`);
+      expect(fs.existsSync(outFile)).toBe(true);
+      expect(fs.readFileSync(outFile, "utf-8")).toContain(`# ${genName}`);
+    });
+  });
 
   describe("raw command (Rust CLI handoff)", () => {
     let projectDir: string;

--- a/packages/turbo-gen/package.json
+++ b/packages/turbo-gen/package.json
@@ -22,10 +22,11 @@
     "test": "jest",
     "check-types": "tsc --noEmit"
   },
-  "dependencies": {},
+  "dependencies": {
+    "@inquirer/prompts": "^7.10.1"
+  },
   "devDependencies": {
     "@arethetypeswrong/cli": "0.18.2",
-    "@inquirer/prompts": "^7.10.1",
     "@jest/globals": "30.2.0",
     "@turbo/test-utils": "workspace:*",
     "@turbo/tsconfig": "workspace:*",

--- a/packages/turbo-gen/src/utils/plop.ts
+++ b/packages/turbo-gen/src/utils/plop.ts
@@ -1,11 +1,14 @@
 import path from "node:path";
+import { createRequire } from "node:module";
 import fs from "fs-extra";
 import type { Project } from "@turbo/workspaces";
 import type { NodePlopAPI, PlopGenerator } from "node-plop";
 import nodePlopModule from "node-plop";
-import { Separator } from "@inquirer/prompts";
+import * as inquirerPrompts from "@inquirer/prompts";
 import { searchUp, getTurboConfigs, logger } from "@turbo/utils";
 import { GeneratorError } from "./error";
+
+const { Separator } = inquirerPrompts;
 
 // Bun's require() of CJS modules with Babel interop wraps exports differently
 const nodePlop = (
@@ -46,59 +49,66 @@ export async function getPlop({
   const workspaceConfigs = getWorkspaceGeneratorConfigs({ project });
   let plop: NodePlopAPI | undefined;
 
-  if (configPath) {
-    if (!fs.existsSync(configPath)) {
-      throw new GeneratorError(`No config at "${configPath}"`, {
-        type: "plop_no_config"
-      });
-    }
-
-    try {
-      plop = await nodePlop(configPath, {
-        destBasePath: configPath,
-        force: false
-      });
-    } catch (e) {
-      logger.error(e);
-    }
-  } else {
-    for (const possiblePath of SUPPORTED_ROOT_GENERATOR_CONFIGS) {
-      const plopFile = path.join(project.paths.root, possiblePath);
-      if (!fs.existsSync(plopFile)) {
-        continue;
-      }
-
-      try {
-        plop = await nodePlop(plopFile, {
-          destBasePath: project.paths.root,
-          force: false
+  try {
+    if (configPath) {
+      if (!fs.existsSync(configPath)) {
+        throw new GeneratorError(`No config at "${configPath}"`, {
+          type: "plop_no_config"
         });
-        break;
-      } catch (e) {
-        logger.error(e);
       }
-    }
 
-    if (!plop && workspaceConfigs.length > 0) {
-      plop = await nodePlop(workspaceConfigs[0].config, {
-        destBasePath: workspaceConfigs[0].root,
-        force: false
-      });
-      workspaceConfigs.shift();
-    }
-  }
-
-  if (plop) {
-    for (const c of workspaceConfigs) {
       try {
-        await plop.load(c.config, {
-          destBasePath: c.root,
+        plop = await nodePlop(await bundleConfigForLoading(configPath), {
+          destBasePath: configPath,
           force: false
         });
       } catch (e) {
         logger.error(e);
       }
+    } else {
+      for (const possiblePath of SUPPORTED_ROOT_GENERATOR_CONFIGS) {
+        const plopFile = path.join(project.paths.root, possiblePath);
+        if (!fs.existsSync(plopFile)) {
+          continue;
+        }
+
+        try {
+          plop = await nodePlop(await bundleConfigForLoading(plopFile), {
+            destBasePath: project.paths.root,
+            force: false
+          });
+          break;
+        } catch (e) {
+          logger.error(e);
+        }
+      }
+
+      if (!plop && workspaceConfigs.length > 0) {
+        plop = await nodePlop(
+          await bundleConfigForLoading(workspaceConfigs[0].config),
+          {
+            destBasePath: workspaceConfigs[0].root,
+            force: false
+          }
+        );
+        workspaceConfigs.shift();
+      }
     }
+
+    if (plop) {
+      for (const c of workspaceConfigs) {
+        try {
+          await plop.load(await bundleConfigForLoading(c.config), {
+            destBasePath: c.root,
+            force: false
+          });
+        } catch (e) {
+          logger.error(e);
+        }
+      }
+    }
+  } finally {
+    cleanupBundledConfigs();
   }
 
   return plop;
@@ -110,7 +120,7 @@ export async function getCustomGenerators({
 }: {
   project: Project;
   configPath?: string;
-}): Promise<Array<Generator | Separator>> {
+}): Promise<Array<Generator | InstanceType<typeof Separator>>> {
   const plop = await getPlop({ project, configPath });
 
   if (!plop) {
@@ -147,7 +157,8 @@ export async function getCustomGenerators({
     }
   });
 
-  const gensWithSeparators: Array<Generator | Separator> = [];
+  const gensWithSeparators: Array<Generator | InstanceType<typeof Separator>> =
+    [];
   Object.keys(gensByWorkspace).forEach((group) => {
     gensWithSeparators.push(new Separator(group));
     gensWithSeparators.push(...gensByWorkspace[group]);
@@ -208,6 +219,124 @@ function injectTurborepoData({
       configs: turboConfigs
     }
   };
+}
+
+// In standalone Bun-compiled binaries, node-plop's require() cannot resolve
+// npm packages from dynamically loaded config files because the binary's module
+// resolution doesn't see the project's node_modules. We use Bun.build() at
+// runtime to bundle the user's config into a single CJS file before node-plop
+// loads it.
+//
+// A subtlety: packages that the binary itself bundles (like @inquirer/prompts)
+// won't exist on disk in the user's project. A Bun.build() plugin intercepts
+// these unresolvable bare specifiers and redirects them to a globalThis
+// registry where the binary's own module references are stored at runtime.
+const bundled = new Set<string>();
+
+// Modules bundled in the compiled binary that user configs may import.
+// These are registered on globalThis before bundling so the generated CJS
+// code can access them without require() (which can't resolve from the
+// binary's virtual filesystem).
+const BINARY_MODULES: Record<string, unknown> = {
+  "@inquirer/prompts": inquirerPrompts
+};
+
+async function bundleConfigForLoading(configPath: string): Promise<string> {
+  const BunAPI = (globalThis as unknown as { Bun?: { build: Function } }).Bun;
+  if (!BunAPI?.build) return configPath;
+
+  try {
+    const outName = path
+      .basename(configPath)
+      .replace(/\.(ts|js|cjs)$/, ".turbo-gen-bundled.cjs");
+    const outDir = path.dirname(configPath);
+    const outPath = path.join(outDir, outName);
+
+    // Expose binary-bundled modules on globalThis so the generated CJS code
+    // can reference them without going through require().
+    const g = globalThis as unknown as Record<string, unknown>;
+    g.__turboGenModules = BINARY_MODULES;
+
+    const result = await BunAPI.build({
+      entrypoints: [configPath],
+      outdir: outDir,
+      naming: outName,
+      target: "bun",
+      format: "cjs",
+      plugins: [binaryResolvePlugin(configPath)]
+    });
+
+    if (
+      !result.success ||
+      !result.outputs ||
+      (result.outputs as Array<unknown>).length === 0
+    ) {
+      return configPath;
+    }
+
+    bundled.add(outPath);
+    return outPath;
+  } catch {
+    return configPath;
+  }
+}
+
+// Bun.build() plugin: when the user's config imports a bare specifier that
+// can't be resolved from disk (the project's node_modules), check if it's a
+// module the binary ships. If so, redirect to a virtual module that reads
+// from the globalThis.__turboGenModules registry at runtime.
+function binaryResolvePlugin(configPath: string) {
+  return {
+    name: "turbo-gen-binary-resolve",
+    setup(build: {
+      onResolve: (
+        opts: { filter: RegExp },
+        cb: (args: {
+          path: string;
+          resolveDir?: string;
+        }) => { path: string; namespace: string } | undefined
+      ) => void;
+      onLoad: (
+        opts: { filter: RegExp; namespace: string },
+        cb: (args: { path: string }) => { contents: string; loader: string }
+      ) => void;
+    }) {
+      build.onResolve({ filter: /^[^./]/ }, (args) => {
+        const resolveDir = args.resolveDir || path.dirname(configPath);
+        try {
+          const localRequire = createRequire(path.join(resolveDir, "noop.js"));
+          localRequire.resolve(args.path);
+          // Found on disk — let Bun bundle it normally
+          return undefined;
+        } catch {
+          // Not on disk — redirect to virtual namespace if the binary has it
+          if (args.path in BINARY_MODULES) {
+            return { path: args.path, namespace: "turbo-gen-builtin" };
+          }
+          return undefined;
+        }
+      });
+
+      build.onLoad(
+        { filter: /.*/, namespace: "turbo-gen-builtin" },
+        (args) => ({
+          contents: `module.exports = globalThis.__turboGenModules[${JSON.stringify(args.path)}];`,
+          loader: "js"
+        })
+      );
+    }
+  };
+}
+
+function cleanupBundledConfigs() {
+  for (const p of bundled) {
+    try {
+      fs.removeSync(p);
+    } catch {
+      // ignore cleanup failures
+    }
+  }
+  bundled.clear();
 }
 
 function getWorkspaceGeneratorConfigs({ project }: { project: Project }) {

